File: 3754-maximum-manhattan-distance-after-k-changes/3754-maximum-manhattan-distance-after-k-changes.java
import java.util.*;

class Solution {
    public int maxDistance(String s, int k) {
        Map<Character, Integer> freq = new HashMap<>();
        int ans = 0;
        
        freq.put('N', 0);
        freq.put('S', 0);
        freq.put('W', 0);
        freq.put('E', 0);
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            freq.put(c, freq.getOrDefault(c, 0) + 1);
            
            int up = freq.get('N');
            int down = freq.get('S');
            int left = freq.get('W');
            int right = freq.get('E');
            int k_left = k;
            
            if (up > down) {
                int val = Math.min(k_left, down);
                up += 2 * val;
                k_left -= val;
            } else {
                int val = Math.min(k_left, up);
                down += 2 * val;
                k_left -= val;
            }
            
            if (left > right) {
                int val = Math.min(k_left, right);
                left += 2 * val;
                k_left -= val;
            } else {
                int val = Math.min(k_left, left);
                right += 2 * val;
                k_left -= val;
            }
            
            ans = Math.max(ans, Math.abs(right - left) + Math.abs(up - down));
        }
        return ans;
    }
}

File: 3754-maximum-manhattan-distance-after-k-changes/3754-maximum-manhattan-distance-after-k-changes.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        # find max when going only in (NE, NW, SE, SW)
        opp_dir = {'N': (0,-1), 'S': (0,1), 'E': (-1,0),'W': (1,0)}
        max_dist = 0
        for dirs in 'NE','NW','SE','SW':
            x = y = 0
            kk = k
            for d in s:
                if d in dirs and kk:
                    kk -= 1
                    x += opp_dir[d][0]
                    y += opp_dir[d][1]
                else:
                    x -= opp_dir[d][0]
                    y -= opp_dir[d][1]
                max_dist = max(max_dist, abs(x) + abs(y))
        return max_dist